lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,241 @@
+        if (blockInPath == leastCommonAnc || blockInPath == &bb ||
+            blockInPath == successor)
+          continue;
+
+        blocksControlDeps[blockInPath].allControlDeps.insert(&bb);
+      }
+    }
+  }
+
+  // Include nested dependencies to the analysis
+  addDepsOfDeps(funcOp);
+
+  // Extract the forward dependencies out of all the control dependencies
+  identifyForwardControlDeps(funcOp);
+}
+
+void dynamatic::ControlDependenceAnalysis::addDepsOfDeps(
+    mlir::func::FuncOp &funcOp) {
+
+  // For each block, consider each of its dependencies (`oneDep`) and move each
+  // of its dependencies into block's
+  for (Block &block : funcOp.getBlocks()) {
+    BlockControlDeps blockControlDeps = blocksControlDeps[&block];
+    for (auto &oneDep : blockControlDeps.allControlDeps) {
+      DenseSet<Block *> &oneDepDeps = blocksControlDeps[oneDep].allControlDeps;
+      for (auto &oneDepDep : oneDepDeps)
+        blocksControlDeps[&block].allControlDeps.insert(oneDepDep);
+    }
+  }
+}
+
+void dynamatic::ControlDependenceAnalysis::identifyForwardControlDeps(
+    mlir::func::FuncOp &funcOp) {
+  Region &funcReg = funcOp.getRegion();
+
+  // Get dominance, post-dominance and loop information
+  DominanceInfo domInfo;
+  PostDominanceInfo postDomInfo;
+  llvm::DominatorTreeBase<Block, false> &domTree = domInfo.getDomTree(&funcReg);
+  CFGLoopInfo li(domTree);
+
+  for (Block &block : funcReg.getBlocks()) {
+
+    // Consider all block's dependencies
+    for (Block *oneDep : blocksControlDeps[&block].allControlDeps) {
+
+      CFGLoop *loop = li.getLoopFor(oneDep);
+
+      // It is a forward control dependency if:
+      // - `oneDep` is not in a loop;
+      // - `oneDep` is not a loop exit or post-dominates `block`;
+      // - `oneDep` is not a latch block.
+      if (!loop || ((!loop->isLoopLatch(oneDep)) &&
+                    (!loop->isLoopExiting(oneDep) ||
+                     postDomInfo.properlyPostDominates(oneDep, &block))))
+        blocksControlDeps[&block].forwardControlDeps.insert(oneDep);
+    }
+  }
+}
+
+std::optional<DenseSet<Block *>>
+dynamatic::ControlDependenceAnalysis::getBlockAllControlDeps(
+    Block *block) const {
+  if (!blocksControlDeps.contains(block))
+    return std::nullopt;
+
+  return blocksControlDeps.lookup(block).allControlDeps;
+}
+
+std::optional<DenseSet<Block *>>
+dynamatic::ControlDependenceAnalysis::getBlockForwardControlDeps(
+    Block *block) const {
+  if (!blocksControlDeps.contains(block))
+    return std::nullopt;
+
+  return blocksControlDeps.lookup(block).forwardControlDeps;
+}
+
+// Return the map of the control dependencies as stored in the class
+ControlDependenceAnalysis::BlockControlDepsMap
+dynamatic::ControlDependenceAnalysis::getAllBlockDeps() const {
+  return blocksControlDeps;
+}
+
+void dynamatic::ControlDependenceAnalysis::printAllBlocksDeps() const {
+
+  LLVM_DEBUG(llvm::dbgs() << "\n*********************************\n\n";
```

Comment:
If you wrap everything inside the outer parenthesis in braces you should get a bit nicer/wider formatting.